### ISSUE TITLE
How to modify path to source files at instrumentation and report generation phases?

### ISSUE THREAD

**Issue by mass85** [istanbuljs/nyc#844]
I am trying to measure coverage of two React web apps in integration tests. Both bundles are built using webpack and babel-plugin-istanbul in separate Docker stages under /root/ directory, then bundles are copied to the final System Under Test Docker image. A separate container is used for executing cucumber/selenium tests. The same test container has an express server with istanbul-middleware running and collecting coverage.

Two problems:
1. When I display /coverage page served by the express server, I would like to see two main directories with sources, one for each web app. Instead I see some mixture of the web apps' subdirectories and some of the files are missing.
2. When I try to display any of the source files I get error similar to: 
```
Error: ENOENT: no such file or directory, open '/root/src/components/fpgaInfo.js'
    at Error (native)
    at Object.fs.openSync (fs.js:642:18)
    at Object.fs.readFileSync (fs.js:510:33)
    at LookupStore.get (/home/tester/express-istanbul/node_modules/istanbul/lib/store/fslookup.js:40:19)
    at HtmlReport.writeDetailPage (/home/tester/express-istanbul/node_modules/istanbul/lib/report/html.js:411:67)
    at Object.render (/home/tester/express-istanbul/node_modules/istanbul-middleware/lib/core.js:211:16)
    at /home/tester/express-istanbul/node_modules/istanbul-middleware/lib/handlers.js:67:14
    at Layer.handle [as handle_request] (/home/tester/express-istanbul/node_modules/express/lib/router/layer.js:95:5)
    at next (/home/tester/express-istanbul/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/home/tester/express-istanbul/node_modules/express/lib/router/route.js:112:3)
```

My questions:
1. Can I affect the way paths are burned into instrumented code?
2. Can I affect the way paths are resolved during HTML report generation?
3. I understand that HTML reporter tries to open each source file in order to generate HTML page with coverage displayed in each source line. Would it be possible to generate such HTML report without access to original sources, but only by using source map burned into instrumented code?

**Comment by JaKXz** [comment]
This is the wrong repository, I think you're looking for: https://github.com/gotwarlost/istanbul-middleware 

I can't tell you if that library is maintained or not. I'm going to close this issue for now. If you have an issue with the underlying instrumentation and assuming you are using the latest versions of the istanbul api, please open an issue in the monorepo.

**Comment by mass85** [comment]
@JaKXz although problems described in my original post were visible when using istanbul-middleware, the root problem is much more general and questions I asked refer to nyc itself, to report merging. Forget my previous post and let me rephrase this issue.

Let's assume we have a project consisting of web application and web server. Code is maintained in a monorepo with an example structure:
├── features
├── packages
│  ├── server
│  │   ├── src
│  │   └── features
│    ├── web-app
│         ├── src
│         └── test
├── test-environment

So both packages have their own unit/component tests with coverage data generated by nyc in `.nyc_output` subdirectories and html reports in `coverage` subdirectories. Besides that there are integration tests defined using cucumber in `features` directory at monorepo root level. These are run in docker environment. What I want to achieve is to **generate a combined coverage report from all types of tests: component, unit and integration**.

And here we come to the problem with paths. Paths to sources are burned into coverage .json files and I'm guessing also in the instrumented web app bundles. So when I collect all coverage .json files from each test run, save it in directory `cov_reports` and generate a report with `nyc report -r html --temp-directory cov_reports/`, the resulting HTML page will contain table with sources in directories as following:

1. home/marcin/project/packages/web-app
2. home/marcin/project/packages/server
3. root/packages/web-app
4. usr/lib/node_modules/project/node_modules/server

Problems with that:
1. There is no combined coverage for each of the components. There are two coverage reports for each of the components instead.
2. When I navigate to a view with coverage for a given source file from the 3. and 4., I get an error: 'Unable to lookup source'. This is because path to that source is not correct for the host machine, it is only correct inside of the docker container.

Is there a way to handle such paths automatically? Or I need to hack around that and i.e. modify coverage data from tests executed in docker container and replace these two prefixes: `usr/lib/node_modules/project/node_modules/` and `root/` with `home/marcin/project/`?

**Comment by ThomasGHenry** [comment]
What did you end up going with?

**Comment by mass85** [comment]
I did the hack and used sed to replace paths in coverage data as described in previous comment.

**Comment by ThomasGHenry** [comment]
I was worried you'd say that. Ugh. 

Thanks!

**Comment by mass85** [comment]
Well, at least it works fine ;) I did not spot any problems.
